### DISCUSSION TITLE
vehicles_list() versus available()

### DISCUSSION THREAD

**Discussion by selsrog** [tdorssers/TeslaPy#138] [Q&A]
I notice that if your Tesla is awake, and you start looping in your application using vehicles_list(), it prevents the car from going to sleep.
As the last_seen() function is giving an error, I'm looking for an alternative. 
Just noticed the available() function, would that work for continuous polling without keeping the car awake?

**Comment by gbizeau** [comment]
This is how I do it.

vehicles = tesla.vehicle_list()

        vehicleData = vehicles[0].get_vehicle_data()

        if vehicleData.get('state') == 'online':

**Reply by selsrog** [reply]
I currently also do it like you describe, but that prevents an online car from going to sleep. And this is something I want to prevent, as it consumes like 1% of the battery...

**Comment by micves** [comment]
It's get_vehicle_data that keeps the car awake. Try something like this:
```
vehicles = tesla.vehicle_list()
vehicle = vehicles[0]
summary=vehicle.get_vehicle_summary()
current_state = summary['state']
print(str(datetime.now()) + ': ' + current_state)

if current_state == "online":
    # do your thing with get_vehicle_data
```
get_vehicle_summary is a light weight info that doesnt keep the car awake. It doesnt contain the same amount of data, but at least you can check if its online.

available() does more or less the same:

```
    def available(self, max_age=60):
        """ Determine vehicle availability based on the cached data or the
        refreshed status when aged out. """
        if self.timestamp + max_age < time.time():
            self.get_vehicle_summary()
        return self['state'] == 'online'
```

**Comment by gbizeau** [comment]
Interesting enough, if you look at whats returned from

tesla.vehicle_list()

It has a state in there... So that's without ditching the vehicle
data. I just changed my code.

**Comment by selsrog** [comment]
Thanks for the feedback. Now, I think it would make sense to give more inside in what I am trying to do.
I want to now when the car is connected to my home charger, so I can make it charge in a controlled way. In the summer for instance I only want solar surplus rolling into it.

Hence, I need to now where it is, and if it is connected at all times. And that is just not possible with vehicle_list, available or the car status. Because it can change with me not checking, and if I check it, I keep the car awake...

Makes sense? Or am I missing something?

Now, I think I found something. last_seen gave me an error (date is in the future), that is why I went looking for another way, as I thought it was a bug. 
But! I just found out that last_seen only gives an error when the car is awake (both on S and 3). But that I can intercept... Moreover, if it gives and error it is online ;-).

So if last_seen is more recent then when I saw the car, I can poll its location and state.  And with that I can work... If it goes offline after me polling, I know it fell asleep without moving...

**Comment by gbizeau** [comment]
If you need to know where it is located, then you are going to have to talk
to the car, hence waking it up to do so. There is no way around that unless
you store it's last location and do some logic to say that it's not online
and hence it's last stored location has to be i's location since it never
came online and therefore could not have moved.... Same goes for the "If
it's connected", I assume you mean plugged in to a charger, only the car
can tell you that...

**Comment by selsrog** [comment]
I understand.
But I was hoping you could fetch a cached version of the car's info online, having a last seen timer.
And this cached info gets updated by the car when it comes online...

That way you would have access to the latest info from the car when he himself was online, without having to wake it up...
But it does seem that this is not available.

Just tested the last_seen(), and that one also keeps the cars awake, if it was awake when when first used in the loop...

**Reply by tdorssers** [reply]
It is possible to fetch cached data when the car is sleeping. The trick is to let the car go to sleep, once it is asleep, you can use `get_vehicle_data()` to retrieve the cached data from Tesla's servers instead of the car. This is also how `last_seen()` should be used.
